### PR TITLE
fix vcat for PooledArray + DataValueArray

### LINKED ISSUE #135: Error with groupby() on DTable
I have a large (~20GB) JuliaDB table saved in 8 chunks that I am now trying to perform some summary functions on. For example, ```:species``` is a column of Strings and I would like to know how many instances I have of each unique string. However, I run into errors when using multiple workers (e.g. ```ERROR: Error deserializing a remote exception from worker 2
Remote(original) exception of type TypeError```). 

Running it on just one worker seems to give me a more informative error:

```julia
using JuliaDB
table = load("output")
result = groupby(length, table, :species)
ERROR: You're using a PooledArray with ref type UInt16, which can only hold 65535 values,
and you just tried to add the 65536th reference.  Please change the ref type
to a larger int type, or use the default ref type (UInt32).
```

I tested out this function on a smaller test table and it seemed to work fine, could there be a problem with the way I have saved or loaded the data? Thanks!

### PR DESCRIPTION
**Pull request by shashi**
fixes #135 

@claireh93 @richardreeve could you confirm?

### CODE REVIEW

**Comment by claireh93** [comment]
Hi @shashi , I have tried this again as a fix for #135 and it seems like `groupby` will now work for one worker, but not several- so that's progress! It gives a slightly vague deserialising error, is there anything else you'd like us to try?

**Comment by shashi** [comment]
Ok, nice! I would like to see the deserialization error.

File: src/util.jl [hunk 1]
```diff
@@ -318,3 +318,6 @@ function tuplesetindex(x::Union{NamedTuple, Tuple}, v::Tuple, i::Tuple)
     reduce((t, j)->tuplesetindex(t, v[j], i[j]), x, 1:length(i))
 end
 
+
+# mild piracy
+Base.similar(::PooledArrays.PooledArray, ::Type{DataValue{T}}, sz::Tuple{Vararg{Int64,N}}) where {T,N} = DataValueArray{T}(sz)
```
Comment (davidanthoff) [inline src/util.jl]: Hehe. I'm starting to wonder whether the method I'm adding to ``similar`` in DataValues.jl is problematic... But if this fixes it for now, I'd call it a day.

**Comment by claireh93** [comment]
Apologies for the delay! The error is:
```julia
Error deserializing a remote exception from worker 8
Remote(original) exception of type TypeError
Remote stacktrace : 
deserialize at /Users/claireh/.julia/v0.6/NamedTuples/src/NamedTuples.jl:402
handle_deserialize at ./serialize.jl:685
deserialize at ./serialize.jl:1079
handle_deserialize at ./serialize.jl:726
deserialize at ./serialize.jl:1079
handle_deserialize at ./serialize.jl:726
deserialize at ./serialize.jl:1008
handle_deserialize at ./serialize.jl:726
deserialize at ./serialize.jl:637
deserialize_datatype at ./serialize.jl:986
handle_deserialize at ./serialize.jl:677
deserialize at ./serialize.jl:637
deserialize_datatype at ./serialize.jl:997
handle_deserialize at ./serialize.jl:677
deserialize at ./serialize.jl:637
deserialize_datatype at ./serialize.jl:997
handle_deserialize at ./serialize.jl:677
deserialize at ./serialize.jl:637
handle_deserialize at ./serialize.jl:689
deserialize at ./serialize.jl:637
#1 at ./serialize.jl:739
ntuple at ./tuple.jl:108
deserialize_tuple at ./serialize.jl:739
handle_deserialize at ./serialize.jl:667
deserialize at ./serialize.jl:637
#1 at ./serialize.jl:739
ntuple at ./tuple.jl:108
deserialize_tuple at ./serialize.jl:739
handle_deserialize at ./serialize.jl:667
deserialize_msg at ./distributed/messages.jl:98
message_handler_loop at ./distributed/process_messages.jl:161
process_tcp_streams at ./distributed/process_messages.jl:118
#99 at ./event.jl:73
#remotecall_fetch#141(::Array{Any,1}, ::Function, ::Function, ::Base.Distributed.Worker, ::Dagger.Context, ::Vararg{Any,N} where N) at ./distributed/remotecall.jl:354
remotecall_fetch(::Function, ::Base.Distributed.Worker, ::Dagger.Context, ::Vararg{Any,N} where N) at ./distributed/remotecall.jl:346
#remotecall_fetch#144(::Array{Any,1}, ::Function, ::Function, ::Int64, ::Dagger.Context, ::Vararg{Any,N} where N) at ./distributed/remotecall.jl:367
macro expansion at /Users/claireh/.julia/v0.6/Dagger/src/compute.jl:332 [inlined]
(::Dagger.##69#70{Dagger.Context,Dagger.OSProc,Int64,Base.#vcat,Tuple{Dagger.Chunk{Array{Tuple{Dagger.Chunk{T,MemPool.DRef} where T,IndexedTables.Columns{NamedTuples._NT_species{String},C} where C<:Union{NamedTuples._NT_, Tuple}},1},MemPool.DRef},Dagger.Chunk{Array{Tuple{Dagger.Chunk{T,MemPool.DRef} where T,IndexedTables.Columns{NamedTuples._NT_species{String},C} where C<:Union{NamedTuples._NT_, Tuple}},1},MemPool.DRef}},Channel{Any},Bool,Bool,Bool})() at ./event.jl:73
```

**Comment by shashi** [comment]
Thanks! Unfortunately that error message is not that helpful :sweat_smile: The line is here https://github.com/blackrock/NamedTuples.jl/blob/c6d1935528a08b10d51147be5d8283fad0283e4f/src/NamedTuples.jl#L398

Did you try loading the data from CSV files all over again? (starting from scratch)

**Comment by einaro23** [comment]
I receive a similar error when I try to merge two JuliaDB tables as shown below:

You're using a PooledArray with ref type UInt8, which can only hold 255 values,
and you just tried to add the 256th reference.  Please change the ref type
to a larger int type, or use the default ref type (UInt32).
unsafe_pool_push!(::PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}}, ::String) at /Users/eotnes/.julia/v0.6/PooledArrays/src/PooledArrays.jl:328
getpoolidx(::PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}}, ::String) at /Users/eotnes/.julia/v0.6/PooledArrays/src/PooledArrays.jl:320
setindex! at /Users/eotnes/.julia/v0.6/PooledArrays/src/PooledArrays.jl:341 [inlined]
copyelt! at /Users/eotnes/.julia/v0.6/IndexedTables/src/columns.jl:387 [inlined]
#55 at /Users/eotnes/.julia/v0.6/IndexedTables/src/columns.jl:395 [inlined]
foreach at /Users/eotnes/.julia/v0.6/IndexedTables/src/utils.jl:52 [inlined]
copyrow!(::IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}}, ::Int64, ::IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}}, ::Int64) at /Users/eotnes/.julia/v0.6/IndexedTables/src/columns.jl:395
_merge!(::IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}}, ::IndexedTables.Columns{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},NamedTuples._NT_Open_High_Low_Close{DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1}}}, ::IndexedTables.NDSparse{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},Tuple{String,Date},IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}},IndexedTables.Columns{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},NamedTuples._NT_Open_High_Low_Close{DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1}}}}, ::IndexedTables.NDSparse{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},Tuple{String,Date},IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}},IndexedTables.Columns{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},NamedTuples._NT_Open_High_Low_Close{DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1}}}}, ::IndexedTables.#right) at /Users/eotnes/.julia/v0.6/IndexedTables/src/join.jl:788
_merge(::IndexedTables.NDSparse{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},Tuple{String,Date},IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}},IndexedTables.Columns{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},NamedTuples._NT_Open_High_Low_Close{DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1}}}}, ::IndexedTables.NDSparse{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},Tuple{String,Date},IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}},IndexedTables.Columns{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},NamedTuples._NT_Open_High_Low_Close{DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1}}}}, ::Function) at /Users/eotnes/.julia/v0.6/IndexedTables/src/join.jl:750
(::Base.#kw##merge)(::Array{Any,1}, ::Base.#merge, ::IndexedTables.NDSparse{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},Tuple{String,Date},IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}},IndexedTables.Columns{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},NamedTuples._NT_Open_High_Low_Close{DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1}}}}, ::IndexedTables.NDSparse{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},Tuple{String,Date},IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}},IndexedTables.Columns{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},NamedTuples._NT_Open_High_Low_Close{DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1}}}}) at ./<missing>:0
_merge(::JuliaDB.#overlap_merge#327{IndexedTables.#right}, ::IndexedTables.NDSparse{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},Tuple{String,Date},IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}},IndexedTables.Columns{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},NamedTuples._NT_Open_High_Low_Close{DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1}}}}, ::IndexedTables.NDSparse{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},Tuple{String,Date},IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}},IndexedTables.Columns{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},NamedTuples._NT_Open_High_Low_Close{DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1}}}}) at /Users/eotnes/.julia/v0.6/JuliaDB/src/ndsparse.jl:214
(::JuliaDB.##326#328{JuliaDB.#overlap_merge#327{IndexedTables.#right}})(::IndexedTables.NDSparse{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},Tuple{String,Date},IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}},IndexedTables.Columns{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},NamedTuples._NT_Open_High_Low_Close{DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1}}}}, ::Vararg{IndexedTables.NDSparse{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},Tuple{String,Date},IndexedTables.Columns{NamedTuples._NT_Ticker_Date{String,Date},NamedTuples._NT_Ticker_Date{PooledArrays.PooledArray{String,UInt8,1,Array{UInt8,1}},Array{Date,1}}},IndexedTables.Columns{NamedTuples._NT_Open_High_Low_Close{DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64},DataValues.DataValue{Float64}},NamedTuples._NT_Open_High_Low_Close{DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1},DataValues.DataValueArray{Float64,1}}}},N} where N) at /Users/eotnes/.julia/v0.6/JuliaDB/src/join.jl:275
do_task(::Dagger.Context, ::Dagger.OSProc, ::Int64, ::Function, ::Tuple{Dagger.Chunk{Any,MemPool.DRef},Dagger.Chunk{Any,MemPool.DRef}}, ::Bool, ::Bool, ::Bool) at /Users/eotnes/.julia/v0.6/Dagger/src/scheduler.jl:254
(::Base.Distributed.##135#136{Dagger.Sch.#do_task,Tuple{Dagger.Context,Dagger.OSProc,Int64,JuliaDB.##326#328{JuliaDB.#overlap_merge#327{IndexedTables.#right}},Tuple{Dagger.Chunk{Any,MemPool.DRef},Dagger.Chunk{Any,MemPool.DRef}},Bool,Bool,Bool},Array{Any,1}})() at ./distributed/remotecall.jl:314
run_work_thunk(::Base.Distributed.##135#136{Dagger.Sch.#do_task,Tuple{Dagger.Context,Dagger.OSProc,Int64,JuliaDB.##326#328{JuliaDB.#overlap_merge#327{IndexedTables.#right}},Tuple{Dagger.Chunk{Any,MemPool.DRef},Dagger.Chunk{Any,MemPool.DRef}},Bool,Bool,Bool},Array{Any,1}}, ::Bool) at ./distributed/process_messages.jl:56
#remotecall_fetch#140(::Array{Any,1}, ::Function, ::Function, ::Base.Distributed.LocalProcess, ::Dagger.Context, ::Vararg{Any,N} where N) at ./distributed/remotecall.jl:339
remotecall_fetch(::Function, ::Base.Distributed.LocalProcess, ::Dagger.Context, ::Vararg{Any,N} where N) at ./distributed/remotecall.jl:339
#remotecall_fetch#144(::Array{Any,1}, ::Function, ::Function, ::Int64, ::Dagger.Context, ::Vararg{Any,N} where N) at ./distributed/remotecall.jl:367
macro expansion at /Users/eotnes/.julia/v0.6/Dagger/src/scheduler.jl:267 [inlined]
(::Dagger.Sch.##13#14{Dagger.Context,Dagger.OSProc,Int64,JuliaDB.##326#328{JuliaDB.#overlap_merge#327{IndexedTables.#right}},Tuple{Dagger.Chunk{Any,MemPool.DRef},Dagger.Chunk{Any,MemPool.DRef}},Channel{Any},Bool,Bool,Bool})() at ./event.jl:73

**Pull request merged**